lib/smack/Prelude.cpp

```diff
@@ -0,0 +1,534 @@
+    s << Decl::axiom(Expr::eq(Expr::id(t.str()),
+      rep.pointerLit((unsigned long long) i)))
+      << "\n";
+  }
+  s << "\n";
+
+  s << "// Memory maps (" << rep.regions->size() << " regions)" << "\n";
+  for (auto M : rep.memoryMaps())
+    s << "var " << M.first << ": " << M.second << ";" << "\n";
+
+  s << "\n";
+
+  s << "// Memory address bounds" << "\n";
+  s << Decl::axiom(Expr::eq(Expr::id(Naming::GLOBALS_BOTTOM),
+    rep.pointerLit(rep.globalsOffset)))
+    << "\n";
+  s << Decl::axiom(
+    Expr::eq(Expr::id(Naming::EXTERNS_BOTTOM),
+      Expr::fn("$add.ref", Expr::id(Naming::GLOBALS_BOTTOM),
+        rep.pointerLit(rep.externsOffset))))
+    << "\n";
+  unsigned long long malloc_top;
+  if (rep.ptrSizeInBits == 32)
+    malloc_top = 2147483647UL;
+  else if (rep.ptrSizeInBits == 64)
+    malloc_top = 9223372036854775807UL;
+  else
+    llvm_unreachable("Unexpected pointer bit width.");
+  s << Decl::axiom(Expr::eq(Expr::id(Naming::MALLOC_TOP),
+    rep.pointerLit(malloc_top)))
+    << "\n";
+  s << "\n";
+
+  if (SmackOptions::MemorySafety) {
+    s << "// Global allocations" << "\n";
+    std::list<const Stmt*> stmts;
+    for (auto E : rep.globalAllocations)
+      stmts.push_back(Stmt::call("$galloc",
+        {rep.expr(E.first), rep.pointerLit(E.second)}));
+    s << Decl::procedure("$global_allocations", {}, {}, {},
+      {Block::block("", stmts)})
+      << "\n";
+    s << "\n";
+  }
+
+  s << "// Bitstd::vector-integer conversions" << "\n";
+  std::string b = std::to_string(rep.ptrSizeInBits);
+  std::string bt = "bv" + b;
+  std::string it = "i" + b;
+  s << Decl::function(rep.indexedName("$int2bv",{rep.ptrSizeInBits}),
+    {{"i",it}}, bt, nullptr, {makeBuiltinAttr("(_ int2bv " + b + ")")})
+    << "\n";
+  if (SmackOptions::BitPrecise) {
+    s << Decl::function(rep.indexedName("$bv2uint",{rep.ptrSizeInBits}),
+      {{"i",bt}}, it, nullptr, {makeBuiltinAttr("bv2int")})
+      << "\n";
+    const Expr* arg = Expr::id("i");
+    const Expr* uint = Expr::fn(rep.indexedName("$bv2uint",
+      {rep.ptrSizeInBits}), arg);
+    std::string offset;
+    if (rep.ptrSizeInBits == 32)
+      offset = "4294967296";
+    else if (rep.ptrSizeInBits == 64)
+      offset = "18446744073709551616";
+    else
+      llvm_unreachable("Unexpected pointer bit width.");
+    s << Decl::function(rep.indexedName("$bv2int",{rep.ptrSizeInBits}),
+      {{"i",bt}}, it, Expr::cond(
+        Expr::fn(rep.indexedName("$slt", {bt, Naming::BOOL_TYPE}),
+          {arg, Expr::lit(0ULL, rep.ptrSizeInBits)}),
+        Expr::fn(rep.indexedName("$sub", {it}), {uint, Expr::lit(offset, 0U)}),
+        uint),
+      {makeInlineAttr()});
+  } else
+    s << Decl::function(rep.indexedName("$bv2int",{rep.ptrSizeInBits}),
+      {{"i",bt}}, it, nullptr, {makeBuiltinAttr("bv2int")})
+      << "\n";
+  s << "\n";
+
+  if (SmackOptions::BitPrecise) {
+    // XXX TODO don't assume 64-bit pointers TODO XXX
+    s << "// Bytewise pointer storage" << "\n";
```

Comment:
We may want these strings to be generated from functions.

Reply:
What would be an appropriate name for such a function?

Reply:
Sorry, I confused you. I meant the function strings below this line.

Reply:
Thanks, I've just made this change. Let me know if it's okay.